<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Using Jupyter (Good), Exercises</span></div>

## Exercise 1 (Learning to interrupt)

In [1]:
while True:
    print('*', end='')

****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

KeyboardInterrupt: 

## Exercise 2 (Resetting & Restarting the Kernal)

In [5]:
print = 'PRINT'

print('Hello World')

TypeError: 'str' object is not callable

In [1]:
print('Hello World')

Hello World


## Exercise 3 (Timing Stuff)

### Verify that both codes give the same result (HINT: print()).

In [26]:
squares = print([i**2 for i in range(10)])

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [25]:
squares = list()
for i in range(10):
    squares.append(i**2)

print(squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


### Use %%timeit to measure the execution time for each example.

In [27]:
%%timeit

squares=10

38.6 ns ± 0.664 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


### Discuss with a friend to understand what each of the numbers in the %%timeit output represents and write down their meanings.

38.6 ns shows the average time it took to execute the code in one loop. 

± 0.664 is the standard deviation, which shows how much the execution times varied across the different runs. A smaller value will indicate that the execution time was consistent across the runs.

7 runs means that the %%timeit command ran the entire set of loops 7 times to calculate the average execution time and standard deviation. 

For 10,000,000 loops each, it means that in each of the 7 runs, the code was executed 10 million times. The %%timeit command determines the number of loops automatically based on how fast the code executes, aiming for a time measurement in the range of a few seconds. 